created the ReadFile function and unit tests

alicjakwie opened the pull request.

alicjakwie committed:
created the ReadFile function and unit tests

alicjakwie committed:
Unit tests for ReadFile

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	m "github.com/googleinterns/step224-2020/hermes/probe/metrics"
+)
+
+func checkSumFromName(fileName string) (string, error) {
```
You only use this in one place, and it's pretty simple, so maybe just inline it?

alicjakwie replied:
inlined it

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	fileName, ok := target.Journal.Filenames[fileID]
+	if !ok {
+		var status m.ExitStatus
+		status = m.FileMissing
```
Can you just replace this with `status := m.FileMissing`, and remove the line above (and similarly in other places you do it)?

leczb replied:
An intermediate variable isn't even required. Just
```
return fmt.Errorf("ReadFile(ID: %d).%v could not read file as the file with the provided ID does not exist", fileID, m.FileMissing)
```

alicjakwie replied:
status now used in different cases

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		return fmt.Errorf("ReadFile(ID: %d).%v could not read file as the file with the provided ID does not exist", fileID, status)
+	}
+	prefix := fileName[0:9]
+	query := &storage.Query{Prefix: prefix}
```
Just define prefix inline.

alicjakwie replied:
changed to how it was handled in create.go

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	prefix := fileName[0:9]
+	query := &storage.Query{Prefix: prefix}
+	start := time.Now()
+	objIter := client.Bucket(bucketName).Objects(ctx, query)
```
does defining the `objIter` do the talking to the network bit, or is that calling `objIter.Next()` (or some awful mix of the two?)

If it doesn't do the talking to the network thing, you probably want to put the `end := time.Now()` later?

alicjakwie replied:
end := time.Now() moved to later

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		return fmt.Errorf("CreateFile check failed: %w", err)
+	}
+	if obj.Name != fileName {
+		fmt.Errorf("The name %v of the file with id %v in the storage system doesn't match the name %v stored in the Journal", obj.Name, fileID, fileID)
```
Do you not need to return this error?

alicjakwie replied:
error returned now

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	target.LatencyMetrics.APICallLatency[m.APIListFiles][m.Success].Metric("hermes_api_latency_s").AddFloat64(end.Sub(start).Seconds())
+	wantChecksum, err := checkSumFromName(fileName)
+	if err != nil {
+		return err
```
You should wrap this error.

alicjakwie replied:
now inlined it

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("The calculated checksum does not match the checksum stored in the file name")
```
Should this have a status bit as well?

alicjakwie replied:
It doesn't for now but it's a TODO for metrics.go and will be changed in another PR

cianbgoog commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		return fmt.Errorf("The calculated checksum does not match the checksum stored in the file name")
+	}
+	if err := reader.Close(); err != nil {
+		return fmt.Errorf("Writer.Close: %v", err)
```
Do we care about the error here? Also, it's `reader.Close()`, not `Writer.Close()`

alicjakwie replied:
changed to defer reader.Close()

cianbgoog commented on hermes/probe/read/read_file_test.go:
```diff
@@ -0,0 +1,80 @@
+	ctx := context.Background()
+	bucketName := "test_bucket_probe0"
+	client := fakegcs.NewClient()
+	fbh := client.Bucket(bucketName)                         // fakeBucketHandle
```
Make the name more descriptive, and delete the comment.

leczb replied:
+1
`bucketHandle` would do.

alicjakwie replied:
changed to fakeBucketHandle comment deleted

cianbgoog commented on hermes/probe/read/read_file_test.go:
```diff
@@ -0,0 +1,80 @@
+	bucketName := "test_bucket_probe0"
+	client := fakegcs.NewClient()
+	fbh := client.Bucket(bucketName)                         // fakeBucketHandle
+	if err := fbh.Create(ctx, bucketName, nil); err != nil { // creates the bucket with name "test_bucket_probe0"
```
We don't need this comment.

alicjakwie replied:
removed

cianbgoog commented on hermes/probe/read/read_file_test.go:
```diff
@@ -0,0 +1,80 @@
+	client := fakegcs.NewClient()
+	fbh := client.Bucket(bucketName)                         // fakeBucketHandle
+	if err := fbh.Create(ctx, bucketName, nil); err != nil { // creates the bucket with name "test_bucket_probe0"
+		t.Error(err)
```
Should this be `t.Fatalf()` with a wrapped error, since this is setup?

alicjakwie replied:
changed to Fatalf for fatalf have to use %v to anotate errors

cianbgoog commented on hermes/probe/read/read_file_test.go:
```diff
@@ -0,0 +1,80 @@
+		t.Error(err)
+	}
+	if err := ReadFile(ctx, target, fileID, client, nil); err != nil {
+		t.Error(err)
```
You should wrap these errors, with some description as to what's wrong (here & elsewhere)

leczb replied:
Meta-comment: "wrapping" has a particular meaning on Go. I think Cian means "annotate", as in, give some context in to the error message:
`t.Errorf("ReadFile() failed: %v", err)`

alicjakwie replied:
Done with %w

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	if len(fileName) < 11 {
+		return "", fmt.Errorf("The filename stored in the journal is shorter than 11 characters, hence does not match Hermes_id_checksum")
+	}
+	return fileName[10:len(fileName)], nil
```
10 is a magic number
https://en.wikipedia.org/wiki/Magic_number_(programming)

Avoid magic numbers in general.
I would put the prefix in a string const then use `len()`.
Make sure that string const is referenced everywhere instead of hard-coding the string in-line.

evanSpendlove replied:
A good use of `len()` here would be: `len("Hermes_ID_"`).

leczb replied:
Even better would be to use the recently introduced FileNamePrefixLength variable.

Also note that there's no need to specify `len(fileName)`, just `return filenName[FileNamePrefxLength:], nil`.

alicjakwie replied:
removed 10 added a const

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	"github.com/google/cloudprober/logger"
+	"github.com/googleapis/google-cloud-go-testing/storage/stiface"
+	"github.com/googleinterns/step224-2020/hermes/probe"
+	m "github.com/googleinterns/step224-2020/hermes/probe/metrics"
```
Don't rename imports for convenience.

alicjakwie replied:
removed

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	if !ok {
+		var status m.ExitStatus
+		status = m.FileMissing
+		return fmt.Errorf("ReadFile(ID: %d).%v could not read file as the file with the provided ID does not exist", fileID, status)
```
"does not exist" is misleading. It's *missing* from the journal. "does not exist" suggests that the storage system doesn't know about it.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		status = m.FileMissing
+		return fmt.Errorf("ReadFile(ID: %d).%v could not read file as the file with the provided ID does not exist", fileID, status)
+	}
+	prefix := fileName[0:9]
```
9 is a magic number. Please fix.

alicjakwie replied:
changed to how it was suggested in create.go

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		return fmt.Errorf("CreateFile check failed: %w", err)
+	}
+	if obj.Name != fileName {
+		fmt.Errorf("The name %v of the file with id %v in the storage system doesn't match the name %v stored in the Journal", obj.Name, fileID, fileID)
```
The error messages says "name" but it inserts the ID instead.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+		fmt.Errorf("The name %v of the file with id %v in the storage system doesn't match the name %v stored in the Journal", obj.Name, fileID, fileID)
+	}
+	target.LatencyMetrics.APICallLatency[m.APIListFiles][m.Success].Metric("hermes_api_latency_s").AddFloat64(end.Sub(start).Seconds())
+	wantChecksum, err := checkSumFromName(fileName)
```
Move this closer to its usage.

Also, follow this pattern if the variable is only used within the `if` block:
```
if want := checkSumFromName(filenName); got != want {
```

alicjakwie replied:
Now getting the checksum from name inline!

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("The calculated checksum does not match the checksum stored in the file name")
```
Nit: error messages should start with lower-case.
Also, include the expected and the calculated checksum in the error.

alicjakwie replied:
Included checksums, all error messages start from lower case now apart from those that start with the function name

leczb commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("The calculated checksum does not match the checksum stored in the file name")
+	}
+	if err := reader.Close(); err != nil {
```
Shouldn't the reader by closed in any case? If so, use `defer`.

alicjakwie replied:
Yes it should. Used defer

deniadelcid commented on hermes/probe/read/read_file.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+func checkSumFromName(fileName string) (string, error) {
+	if len(fileName) < 11 {
```
For the average reader the it's not trivial to see where this `11` came from. Instead of hardcoding it, can you instead make it a constant? e.g. filePrefixSize := len("Hermes_XX_") 

As a general tip, use constants and/or flags instead of hardcoding.

alicjakwie replied:
this function has now been removed and the verification is done inline think the process is much cleaner :)

leczb reviewed (commented):
This PR shows up as requiring review from me, but I've already reviewed it.
Please request review again when it's ready for another round.

alicjakwie committed:
Update and rename read_file.go to read.go

alicjakwie committed:
Update and rename read_file_test.go to read_test.go

alicjakwie committed:
Update read_test.go

alicjakwie commented:
Addressed all the comments in the PR, applied changes suggested in the CreateFile PR, changed the name of the files (from read_file.go to read.go) to read.go to match the package name, added doc strings and a license header, left #76 as a TODO.

cianbgoog commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
 		TotalSpaceAllocatedMib: int64(100),
```
If our filesize is 100 bytes, why are we allocating 100MiB?

alicjakwie replied:
Thank you for catching that! I chaned to 1MiB I miscalculated....

cianbgoog commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
+	bucketName := "test_bucket_probe0"
+	fakeBucketHandle := client.Bucket(bucketName)
+	if err := fakeBucketHandle.Create(ctx, bucketName, nil); err != nil {
 		t.Error(err)
```
Wrap this error, so `t.Errorf("Error creating bucket: %w", err)`

alicjakwie replied:
Error Anoted!

cianbgoog reviewed (approved):
minor comments, feel free to merge when you've made changes to address.

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+)
+
+const (
+	FileNameFormat          = "Hermes_%02d_%x"
```
Exported symbols should be documented.

alicjakwie replied:
documentation added

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+	minFileID               = 1
+	maxFileID               = 50
+	maxFileSizeBytes        = 1000
+	hermesAPILatencySeconds = "hermes_api_latency_s"
```
Isn't this in the `metrics` package now?

deniadelcid replied:
That change hasn't been merged yet.
https://github.com/googleinterns/step224-2020/blob/master/hermes/probe/metrics/metrics.go

alicjakwie replied:
It is but not as a constant as in (changing this now it's) "hermes_api_latency_seconds"

alicjakwie replied:
left it as a constant will be removed once the metrics.go PR is merged

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+		return fmt.Errorf("file with the ID: %d is missing from the State Journal", fileID)
+	}
+
+	// Verify that the file that we want to read is in fact present in the target system
```
You could factor out this part of the function into another function named `verifyFileExists` (or something to that effect). This would obviate the need for this comment and also make the function shorter and more readable.

alicjakwie replied:
function factored out!

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+	finish := time.Now()
+	if len(namesFound) == 0 {
+		target.LatencyMetrics.APICallLatency[metrics.APIListFiles][metrics.FileMissing].Metric(hermesAPILatencySeconds).AddFloat64(finish.Sub(start).Seconds())
+		return fmt.Errorf("ReadFile(ID: %d) could not read file as the file with the provided ID does not exist on the target system", fileID)
```
Maybe log the bucket name as well to aid debugging?

alicjakwie replied:
Added

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+		return fmt.Errorf("ReadFile(ID: %d) could not read file as the file with the provided ID does not exist on the target system", fileID)
+	}
+	if len(namesFound) != 1 {
+		return fmt.Errorf("expected exactly one file in bucket %q with prefix %q; found %d: %v", bucketName, fileNamePrefix, len(namesFound), namesFound)
```
Keep the error messages consistent: either put the function name in all of them or none of them.
I would lean towards not putting the function name there, as it should be the responsibility of the caller to include it (if it's relevant).

alicjakwie replied:
removed function name from errors added in in read_test.go errors starting from a small letter

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+	defer reader.Close()
+	h := sha1.New()
+	if _, err := io.Copy(h, reader); err != nil {
+		return fmt.Errorf("io.Copy: %v", err)
```
Maybe add a prefix to the error message saying that the checksum calculation failed.

alicjakwie replied:
added

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+		return fmt.Errorf("io.Copy: %v", err)
+	}
+	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
+	wantChecksum := fileName[len(fileNamePrefix):len(fileName)]
```
Nit: There's no need for the second `len()`. Just `wantChecksum := fileName[len(fileNamePrefix):]`.

alicjakwie replied:
changed!

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
+	wantChecksum := fileName[len(fileNamePrefix):len(fileName)]
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("the calculated checksum: %q does not match the checksum in the file name: %q", gotChecksum, wantChecksum)
```
Shouldn't this finding be recorded in the metric?

alicjakwie replied:
This is a TODO for metrics (more exit statuses need to be added)

leczb commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("the calculated checksum: %q does not match the checksum in the file name: %q", gotChecksum, wantChecksum)
+	}
+	logger.Infof("verified consiistency for object %q in bucket %q.", fileName, bucketName)
```
Typo.
Also, there's no need for the period at the end.

alicjakwie replied:
fixed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
+	ctx := context.Background()
+	client := fakegcs.NewClient()
+	bucketName := "test_bucket_probe0"
+	fakeBucketHandle := client.Bucket(bucketName)
```
Just ` bucketHandle`. The reader knows it's a fake.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
+	bucketName := "test_bucket_probe0"
+	fakeBucketHandle := client.Bucket(bucketName)
+	if err := fakeBucketHandle.Create(ctx, bucketName, nil); err != nil {
 		t.Error(err)
```
This should be Fatalf, as it's a set-up error.

alicjakwie replied:
changed to Fatalf

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
+		{6, 0, true},
+	}
+	for _, tc := range tests {
+		if err = create.CreateFile(ctx, target, tc.fileIDCreate, fileSizeBytes, client, logger); err != nil {
```
This should be `if err := ...`, not `if err = ...`.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -65,16 +78,44 @@ func TestReadFile(t *testing.T) {
+		if err = create.CreateFile(ctx, target, tc.fileIDCreate, fileSizeBytes, client, logger); err != nil {
+			t.Fatalf("CreateFile(fileID: %d) set up failed %v", tc.fileIDCreate, err)
+		}
+		err = ReadFile(ctx, target, tc.fileIDRead, fileSizeBytes, client, logger)
```
This can be simplified:

```
if err := ...; err != tc.wantErr {
  t.Errorf("ReadFile(fileID: %d) = %w; want %v", tc.fileIDRead, err, tc.wantErr)
}
```

alicjakwie replied:
tc.wantErr is a bool so I am not sure whether this applies?

leczb replied:
My bad. Let's fix it:
```
if err := ...; (err != nil) != tc.wantErr {
  t.Errorf("ReadFile(fileID: %d) = %w; want %v", tc.fileIDRead, err, tc.wantErr)
}
```

alicjakwie replied:
simplified as per suggestion!

deniadelcid commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+//
+// Author: Alicja Kwiecinska, GitHub: alicjakwie
+//
+// package read contains all of the logic necessary to verify the availability and consistency of the file contents and names in GCS.
```
```suggestion
// Package read contains all of the logic necessary to verify the availability and consistency of the file contents and names in GCS.
```

alicjakwie replied:
changed

alicjakwie replied:
changed

deniadelcid commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,124 @@
+//          target: contains information about target storage system, carries an intent log in the form of a StateJournal and it used to export metrics.
+//          fileID: the unique identifer of every file, it cannot be repeated. It needs to be in the range [minFileID, maxFileID]. FileID 0 is reserved for a special file called the NIL file.
+//          client: is a storage client. It is used as an interface to interact with the target storage system.
+//          logger: a cloudprober logger used to record the exit status of the CreateFile operation in a target bucket. The logger passed MUST be a valid logger.
```
Nit: I think that in here you meant to say "ReadFile" instead of "CreateFile" ;)

alicjakwie replied:
changed

deniadelcid commented on hermes/probe/read/read_test.go:
```diff
@@ -0,0 +1,121 @@
+		&probepb.Target{
+			Name:                   "hermes",
+			TargetSystem:           probepb.Target_GOOGLE_CLOUD_STORAGE,
+			TotalSpaceAllocatedMib: int64(1000),
```
In here to you mean Mib (Mebibit) or MiB (Mebibyte)? Usually storage information is represented in bytes not bits. If you meant bytes change this to `TotalSpaceAllocatedMiB`.

Since it appears that this is in another file, you can just open an issue and add a to-do.

alicjakwie replied:
added a TODO: #79

alicjakwie committed:
Update read.go

alicjakwie committed:
Update read_test.go

alicjakwie committed:
Update read.go

alicjakwie committed:
Update read_test.go

alicjakwie committed:
Update read.go

alicjakwie committed:
Update read_test.go

alicjakwie committed:
Update read_test.go

leczb commented on hermes/probe/read/read.go:
```diff
@@ -35,37 +36,15 @@ import (
-	}
-
-	// Verify that the file that we want to read is in fact present in the target system
+func verifyFileExists(ctx context.Context, client stiface.Client, bucketName string, target *probe.Target, fileName string, fileID int32) error {
```
Extra-nit: customarily, the exported symbols come first in the file.

leczb commented on hermes/probe/read/read.go:
```diff
@@ -106,19 +111,20 @@ func ReadFile(ctx context.Context, target *probe.Target, fileID int32, fileSize
-		target.LatencyMetrics.APICallLatency[metrics.APIGetFile][status].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
-		return fmt.Errorf("readFile(id: %d).%v: could not read file %s: %w", fileID, status, fileName, err)
+		target.LatencyMetrics.APICallLatency[metrics.APIGetFile][status].Metric(hermesAPILatencySeconds).AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf(".%q: could not read file %q: %w", status, fileName, err)
```
Nit: I wouldn't use %q to format the status. Also, stray dot?

alicjakwie replied:
changed to %v

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -1,26 +1,11 @@
-// Copyright 2020 Google LLC
```
I assume all source files need the licence header. Is that not so?

alicjakwie replied:
added

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -32,15 +17,16 @@ import (
 			Name:                   "hermes",
 			TargetSystem:           probepb.Target_GOOGLE_CLOUD_STORAGE,
-			TotalSpaceAllocatedMib: int64(1000),
+			TotalSpaceAllocatedMib: int64(1),
```
Is the `int()` required? Wouldn't just `1` work?

alicjakwie replied:
it works!! so changed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -32,15 +17,16 @@ import (
 			Name:                   "hermes",
 			TargetSystem:           probepb.Target_GOOGLE_CLOUD_STORAGE,
-			TotalSpaceAllocatedMib: int64(1000),
+			TotalSpaceAllocatedMib: int64(1),
```
Nit: s/Mib/MiB/

alicjakwie replied:
Added this as a TODO(#79 )

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -54,7 +40,7 @@ func TestReadFile(t *testing.T) {
 				Name:                   "hermes",
 				TargetSystem:           probepb.Target_GOOGLE_CLOUD_STORAGE,
-				TotalSpaceAllocatedMib: int64(100),
+				TotalSpaceAllocatedMib: int64(1),
```
Same as above.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -87,12 +73,14 @@ func TestReadFile(t *testing.T) {
 	ctx := context.Background()
 	client := fakegcs.NewClient()
 	bucketName := "test_bucket_probe0"
```
Nit: this is called `bucket` in other places. I would recommend making it consistent.

alicjakwie replied:
changed

leczb commented on hermes/probe/read/read_test.go:
```diff
@@ -106,16 +94,11 @@ func TestReadFile(t *testing.T) {
-		if err != nil && !tc.wantErr {
-			t.Errorf("ReadFile(fileID: %d) = %w, expected: nil", tc.fileIDRead, err)
+		if err := ReadFile(ctx, target, tc.fileIDRead, fileSizeBytes, client, logger); (err != nil) != tc.wantErr {
+			t.Errorf("ReadFile(fileID: %d) = %w, expected: %v", tc.fileIDRead, err, tc.wantErr)
```
Nit: s/expected/want/

alicjakwie replied:
changed

leczb reviewed (approved):
LGTM

deniadelcid commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,130 @@
+	hermesAPILatencySeconds = "hermes_api_latency_seconds"
+)
+
+func verifyFileExists(ctx context.Context, client stiface.Client, bucketName string, target *probe.Target, fileName string, fileID int32) error {
```
Given that target already contains `bucketName` field, you don't need to pass it as a separate argumeent.

alicjakwie replied:
changed

deniadelcid commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,130 @@
+	if !ok {
+		return fmt.Errorf("file with the ID: %d is missing from the State Journal", fileID)
+	}
+	verifyFileExists(ctx, client, bucketName, target, fileName, fileID)
```
`verifyFileEists` returns an error. Can you add a check to verify the error is nil?

alicjakwie replied:
Done!

evanSpendlove commented on hermes/probe/read/read.go:
```diff
@@ -0,0 +1,130 @@
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("the calculated checksum: %q does not match the checksum in the file name: %q", gotChecksum, wantChecksum)
+	}
+	logger.Infof("verified consistency for object %q in bucket %q", fileName, bucketName)
```
Aren't we meant to re-create the file contents using the random byte generator again and check this checksum against the file name and contents read checksum?

evanSpendlove replied:
Added an issue for this:
https://github.com/googleinterns/step224-2020/issues/80

alicjakwie replied:
That could be an extra feature added in the future!

alicjakwie committed:
Update read.go

alicjakwie committed:
Update read.go

alicjakwie committed:
Update read_test.go

alicjakwie committed:
Update read.go

alicjakwie commented:
All comments addressed! Got all approvals!
